Step2: 지하철 노선 기능 리뷰 요청 드립니다.

programmer-sjk opened the pull request:
안녕하세요 지민님 😄 
2단계 리뷰 부탁드립니다.

코드 리뷰에 도움이 되도록 **특이사항**을 말씀드리면
1. 지하철 노선이 추가되면서 인수테스트가 **개발자가 아닌 다른 사람이 이해하기 어렵다는 느낌**을 받았습니다.
-> 그래서 Station, Line CRUD 함수 이름을 한글로 수정했습니다.
2. `인수 조건` 에는 없었지만 **노선 조회 시, 역 정보도 같이 조회**된다는 인수 테스트를 추가했습니다. 
3.  업데이트 할 때, 컨트롤러에서 `PutMapping` 를 쓸지 `PatchMapping` 를 쓸지 고민했는데요. 제가 이해한 대로면 
`PUT`은 전체 엔티티를 업데이트 하고, 일부(Partial)를 업데이트 할 경우 `Patch` 메소드를 쓴다고 알고 있어서 @PatchMapping을 사용했습니다.

그럼 리뷰 부탁드릴게요 😄

programmer-sjk committed:
docs(guide): README.md 요구사항 및 기능 구현 목록 작성

programmer-sjk committed:
feat(domain): Line domain, repository 추가

programmer-sjk committed:
test(acceptance): 지하철 노선 생성 인수 테스트 작성, 기능 구현

programmer-sjk committed:
test(acceptance): 지하철 노선 목록 조회 인수 테스트 작성

programmer-sjk committed:
test(acceptance): 지하철 노선 조회 인수 테스트 작성

programmer-sjk committed:
feat: 지하철 노선 조회 기능 구현

programmer-sjk committed:
test(acceptance): 지하철 노선 수정 인수 테스트 작성

programmer-sjk committed:
feat: 지하철 노선 수정 기능 개발

programmer-sjk committed:
test(acceptance): 지하철 노선 삭제 인수 테스트 작성

programmer-sjk committed:
feat: 지하철 노선 삭제 기능 구현

programmer-sjk committed:
test: 인수 테스트 격리

programmer-sjk committed:
refactor: 전반적인 코드 리팩토링

리팩토링
- repository 패키지 추가
- 사용되지 않는 함수 제거
- nullable 옵션 추가
- 변수명 수정

programmer-sjk committed:
test(acceptance): 지하철 노선 조회 시, 포함된 지하철 역 조회 인수 테스트 추가

programmer-sjk committed:
test(acceptance): 인수 테스트 가독성을 위해 CRUD 역할을 하는 함수 이름을 한글로 수정

programmer-sjk committed:
test(acceptance): API spec에 맞게 일부 코드 수정

programmer-sjk committed:
refactor: 변수명 수정

programmer-sjk committed:
refactor(dto): Arrays.asList 인자 가독성 좋게 수정

joojimin commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+@Service
+@Transactional(readOnly = true)
+public class LineService {
+    private LineRepository lineRepository;
```
여기도 `final` 키워드가 추가되면 좋을거같아요!

joojimin commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+@Transactional(readOnly = true)
+public class LineService {
+    private LineRepository lineRepository;
+    private final StationService stationService;
```
같은 서비스 레이어끼리의 참조는 순환 참조등을 야기시킬 수 있을 것 같아요!
공통적인 기능을 component로 분리하거나, repo를 주입받아 처리하는건 어떨까요?

joojimin commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,55 @@
+        return ResponseEntity.created(URI.create("/lines/" + line.getId())).body(line);
+    }
+
+    @PatchMapping("/lines/{id}")
```
`3. 업데이트 할 때, 컨트롤러에서 PutMapping 를 쓸지 PatchMapping 를 쓸지 고민했는데요. 제가 이해한 대로면 PUT은 전체 엔티티를 업데이트 하고, 일부(Partial)를 업데이트 할 경우 Patch 메소드를 쓴다고 알고 있어서 @PatchMapping을 사용했습니다.`
=> 저는 애매할경우 POST를 주로 쓰는편인데, 세세하게 고민하신 것 같아요! 저도 말씀해주셔서 찾다보니 Patch 사용에 주의할 점이 있다고합니다! 참고만해주시면될거같아요! 👍 
https://repo.yona.io/doortts/blog/issue/12

joojimin commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,57 @@
+
+    protected Line() {}
+
+    public Line(String name, String color, Station upStation, Station downStation, int distance) {
```
생성자를 통해 3개이상의 인자를 받는경우 `builder 패턴`을 추천하고 있어요! 인자끼리 Type이 같아 잘못 넘길 수 도 있는 실수도 방지할 수 있습니다!

joojimin commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,55 @@
+
+@RestController
+public class LineController {
+    private LineService lineService;
```
변경을 방지하는 키워드가 추가되면 좋을 것 같아요!

joojimin commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity deleteStation(@PathVariable Long id) {
```
제네릭 사용시 타입은 꼭 명시해주는게 좋을 거 같아요!

joojimin commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,263 @@
+            Long downStationId,
+            int distance
+    ) {
+        Map<String, String> params = new HashMap<>();
```
Map보다 만들어뒀던 Request 그대로 쓰는것이 나중에 수정에도 용이하지 않을까요~?

joojimin reviewed (changes requested):
안녕하세요~! 정국님!
미션진행 고생많으셨습니다!
몇가지 확인하고 넘어가면 좋을법만 코멘트 드렸습니다
확인해보시고 리뷰 재요청 부탁드려요~!

programmer-sjk committed:
refactor: 코드 리뷰 반영

코드 리뷰 반영
- final 키워드 추가
- 순환 참조를 피해 repository 주입
- builder 패턴 적용
- Void 제너릭 타입 사용

programmer-sjk commented:
안녕하세요 지민님
코드 리뷰 반영하였습니다.

특이사항 정리해보면
```
같은 서비스 레이어끼리의 참조는 순환 참조등을 야기시킬 수 있을 것 같아요!
공통적인 기능을 component로 분리하거나, repo를 주입받아 처리하는건 어떨까요?
```
-> 리뷰 덕분에 **Service 간 순환 참조**에 대해 공부해 볼 수 있었습니다. **간단한 조회 쿼리는 repository를 주입**받아 사용해도 좋다는 정보들을 근거로 repository 주입 받아 사용하였습니다.

```
생성자를 통해 3개이상의 인자를 받는경우 builder 패턴을 추천하고 있어요! 인자끼리 Type이 같아 잘못 넘길 수 도 있는 실수도 방지할 수 있습니다!
```
-> 이런 패턴이 있고 공부할 수 있어서 너무 좋았습니다!! `Spring`에서 `builder` 패턴으로 검색해보면 많은 내용이 `lombok`의 `Builder 어노테이션`에 대해 다루고 있었는데요. lombok을 사용하려면 gradle에 종속성을 추가해야 하더라구요. 종속성까지 추가하면서 Builder 어노테이션을 추가하는게 지민님 의도였는지 확신이 없어, `내부 static Builder 클래스`를 추가해서 **3개 이상의 매개변수를 가진 domain, dto**를 수정했습니다. 혹시 의도하신게 lombok Builder 어노테이션이였다면 말씀해주세요!

programmer-sjk committed:
refactor: dto 생성자 접근제한자 private로 변경

joojimin reviewed (approved):
피드백 반영 고생하셨습니다 정국님~!!
오 제 피드백은 `builder 패턴`을 활용하기만 하면 되는 것이 였어요!
lombok 종속성을 추가하라는건 아니였습니다 ㅎㅎㅎ

물론 실제 현업에서는 작성해야하는 코드양을 줄이기위해 lombok을 적극사용하고 있지만, 지금 과정은 공부하는 과정이라 거기까지 생각할 필요는 없을 것 같아요!!
미션진행 고생하셨습니다 ㅎㅎㅎ
